### PR TITLE
Add DeoptSafepoint instruction 

### PR DESCRIPTION
**Pull request by jacob-hughes**
This solves the issue where deserialisation and handling of calls from within the yk runtime were special-cased to handle LLVM's stack-map intrinsic. This was brittle and created complex nested pattern matching.

Now, calls to LLVM's stack-map intrinsics are serialised to disk as `DeoptSafepoint` instructions, where the live variables are easily retrievable, and useless duplication (such as the function name) is removed.

I think naming this instruction `Stackmap` is a misnomer because a stack-map is just the name of the data structure on disk. I've named it `DeoptSafepoint` because what this instruction really represents is a call which says "at this moment in time, I know the live variables". In other VMs this is usually known as a safepoint -- at least in the GC world! I personally think this fits better, but I don't mind changing it as you know the system better than I do.

### CODE REVIEW

File: ykrt/src/compile/jitc_yk/aot_ir.rs [hunk 1]
```diff
@@ -406,6 +405,16 @@ pub(crate) enum Instruction {
         binop: BinOp,
         rhs: Operand,
     },
+    #[deku(id = "12")]
```
[the remark this answers is not in the document]
Comment (vext01) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:408]: BTW deku does not warn of re-using the same ID! I should raise an issue.

File: ykrt/src/compile/jitc_yk/aot_ir.rs [hunk 2]
```diff
@@ -406,6 +405,16 @@ pub(crate) enum Instruction {
...
+    DeoptSafepoint {
+        id: Operand,
+        // FIXME: this is never used, do we need to serialize this?
+        num_shadow_bytes: Operand,
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/aot_ir.rs:412]: kill it.

**Comment by vext01** [comment]
LGTM with minor comments.

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 2 above)
Comment (ptersilie) [inline ykrt/src/compile/jitc_yk/aot_ir.rs:412]: No, we won't need this. We can get rid of it. But given we will merge deopt into the call/br instructions they are attached to soon, there's probably no need to do this now.

**Comment by ptersilie** [comment]
Great, this makes the stackmap stuff much more readable.

We need to squash away that first commit I think cause otherwise git bisect won't work past this point. But there's a comment from Edd that we need to fix first.

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 2 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:412]: No, we won't need this. We can get rid of it. But given we will merge deopt into the call/br instructions they are attached to soon, there's probably no need to do this now.

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:408]: I'll bump this to 13 and push an update once you merge your PR.

**Comment by jacob-hughes** [comment]
Thanks. I don't think I can squash this down just yet anyway. I have to wait until my ykllvm PR merges first.

**Comment by ltratt** [comment]
@jacob-hughes You can merge in your PR (though I don't think force pushing is right yet, as you need to make some code changes right?).

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 2 above)
Comment (jacob-hughes) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:412]: I will remove this as part of our new changes because removing this would require a new PR to ykllvm.

**Comment by jacob-hughes** [comment]
I'm ready to squash. I need to bump the deku ID but I need to do that as part of the squash because it is a merge conflict with the base branch. Happy for me to do so?

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 2 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/aot_ir.rs:412]: Works for me.

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**